dns_sd_avahi.c

```diff
@@ -61,82 +211,279 @@ static void __avahi_browser_cb(AvahiServiceBrowser *browser,
 {
 	struct avahi_discovery_data *ddata = (struct avahi_discovery_data *) d;
 	struct AvahiClient *client = avahi_service_browser_get_client(browser);
+	int i;
+
+	if (!browser) {
+		ERROR("Fatal Error in Avahi Browser\n");
+		return;
+	}
 
 	switch (event) {
-	default:
+	case AVAHI_BROWSER_REMOVE:
+		DEBUG("Avahi Browser : REMOVE : "
+				"service '%s' of type '%s' in domain '%s'\n",
+				name, type, domain);
+		break;
 	case AVAHI_BROWSER_NEW:
-		ddata->found = !!avahi_service_resolver_new(client, iface,
+		DEBUG("Avahi Browser : NEW: "
+				"service '%s' of type '%s' in domain '%s'\n",
+				name, type, domain);
+		if(!avahi_service_resolver_new(client, iface,
 				proto, name, type, domain,
 				AVAHI_PROTO_UNSPEC, 0,
-				__avahi_resolver_cb, d);
+				__avahi_resolver_cb, d)) {
+			ERROR("Failed to resolve service '%s\n", name);
+		} else
+			ddata->found++;
 		break;
 	case AVAHI_BROWSER_ALL_FOR_NOW:
-		if (ddata->found) {
-			while (!ddata->resolved) {
-				struct timespec ts;
-				ts.tv_sec = 0;
-				ts.tv_nsec = 4000000;
-				nanosleep(&ts, NULL);
-			}
+		/* Wait for a max of 1 second */
+		i = 0;
+		DEBUG("Avahi Browser : ALL_FOR_NOW Browser : %d, Resolved : %d\n",
+				ddata->found, ddata->resolved);
+		while ((ddata->found != ddata->resolved)  && i <= 200) {
+			struct timespec ts;
+			ts.tv_sec = 0;
+			ts.tv_nsec = 5e6;	/* 5ms in ns*/
+			nanosleep(&ts, NULL);
+			i++;
 		}
-		/* fall-through */
+		avahi_simple_poll_quit(ddata->poll);
+		break;
 	case AVAHI_BROWSER_FAILURE:
+		DEBUG("Avahi Browser : FAILURE\n");
 		avahi_simple_poll_quit(ddata->poll);
-		/* fall-through */
+		break;
 	case AVAHI_BROWSER_CACHE_EXHAUSTED:
+		DEBUG("Avahi Browser : CACHE_EXHAUSTED\n");
 		break;
 	}
 }
 
-int discover_host(char *addr_str, size_t addr_len, uint16_t *port)
+/*
+ * This creates the linked lists, tests it (make sure a context is there)
+ * and returns things. the returned structure must be freed with 
+ * free_all_discovery_data();
+ */
+
+static struct avahi_discovery_data * dnssd_find_hosts(int *ret)
 {
-	struct avahi_discovery_data ddata;
-	int ret = 0;
-	AvahiAddress address;
+	struct avahi_discovery_data *ddata, *ndata;
 	AvahiClient *client;
 	AvahiServiceBrowser *browser;
-	AvahiSimplePoll *poll = avahi_simple_poll_new();
+	int i;
 
-	memset(&address, 0, sizeof(address));
-	if (!poll)
-		return -ENOMEM;
+	ddata = new_discovery_data();
+
+	if (!ddata) {
+		*ret = -ENOMEM;
+		return NULL;
+	}
+
+	ddata->poll = avahi_simple_poll_new();
+	if (!ddata->poll) {
+		*ret = -ENOMEM;
+		return NULL;
+	}
 
-	client = avahi_client_new(avahi_simple_poll_get(poll),
-			0, NULL, NULL, &ret);
+	client = avahi_client_new(avahi_simple_poll_get(ddata->poll),
```

Comment:
this change with `poll`  -> `ddata->poll`  makes things a bit hard to follow; and looks un-needed;
maybe some of this stuff could be moved into `new_discovery_data()` before  adding new logic?

Reply:
The issue is - the poll needs to be kept track of, and passed to __avahi_browser_cb, so that it's in the linked list...

What was happening before - was things were allocated, and the poll and ddata->poll were pointing to the same object, this just removes that (what I considered) obfuscation.

Your point of breaking this up into multiple commits is spot on through...